src/mscorlib/shared/System/Text/StringBuilder.cs

```diff
@@ -790,6 +749,75 @@ public StringBuilder Append(string value, int startIndex, int count)
             }
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                return AppendCore(value, 0, value.Length);
+            }
+            return this;
+        }
+
+        public StringBuilder Append(StringBuilder value, int startIndex, int count)
+        {
+            if (startIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_GenericPositive);
+            }
+
+            if (value == null)
+            {
+                if (startIndex == 0 && count == 0)
+                {
+                    return this;
+                }
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (count > value.Length - startIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
+            }
+
+            return AppendCore(value, startIndex, count);
+        }
+
+        private StringBuilder AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            if (value == this && count > 0)
+                return Append(value.ToString(startIndex, count));
+
+            int newLength = Length + count;
+            Debug.Assert(value != this);
```

Comment:
seems you shouldn't have this assert anymore as you allow value of this. The assert blocks appending yourself with zero count, which should work (no-op)

should be a test for that..

Reply:
```
if (count == 0)
{
     return this;
}
```
I added this check similar to string overloads

Reply:
would this work
```c#
+        private StringBuilder AppendCore(StringBuilder value, int startIndex, int count)
 +        {
 +            if (value == this)
 +                return Append(value.ToString(startIndex, count));
```
ie, simply stop checking for 0.